### PR TITLE
Rem 203

### PR DESCRIPTION
**Pull request by Alinadimmokdad**
strategies section refactoring

### CODE REVIEW

File: src/pages/strategies/forms/indicatorForm.js [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    })
+      .then(res => {
+        const gridData = res.list.map((item, index) => ({
+          id: index + 1,
+          seqNo: item.seqNo,
+          name: item.name,
+          indicatorName: item.indicatorName,
+          indicatorId: item.indicatorId,
+          codeId: item.codeId,
+          strtategyId: item.strategyId
```
Comment (NoraBitarSM) [inline src/pages/strategies/forms/indicatorForm.js:100]: are all these necessary or can be replaces by ...rest ?

File: src/pages/strategies/forms/strategiesForm.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  const { recordId } = store
+
+  const invalidate = useInvalidate({
+    endpointId: DocumentReleaseRepository.Strategy.qry
+  })
+
+  const [initialValues, setInitialData] = useState({
+    recordId: null,
+    name: '',
+    group: '',
```
Comment (NoraBitarSM) [inline src/pages/strategies/forms/strategiesForm.js:30]: do we need group ?  groupId is enough

File: src/pages/strategies/forms/strategiesForm.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    type: '',
+    groupId: ''
+  })
+
+  const strategiesFormik = useFormik({
+    enableReinitialize: true,
+    validateOnChange: true,
+    initialValues,
+    validationSchema: yup.object({
+      name: yup.string().required('This field is required'),
```
Comment (NoraBitarSM) [inline src/pages/strategies/forms/strategiesForm.js:40]: no need for all the 'This field is required' messages

File: src/pages/strategies/forms/strategiesForm.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    }
+  }
+
+  useEffect(() => {
+    if (recordId) {
+      getGroupId(recordId)
+    }
+  }, [recordId, getRequest])
+
+  const getGroupId = recordId => {
```
Comment (NoraBitarSM) [inline src/pages/strategies/forms/strategiesForm.js:92]: getGroupId function to get the strategy record and assign it in the formik instance
the naming is wrong.

File: src/pages/strategies/forms/strategiesForm.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            parameters='_startAt=0&_pageSize=100'
+            name='groupId'
+            label={labels.groupStrat}
+            valueField='recordId'
+            displayField='name'
+            values={strategiesFormik.values}
+            required
+            readOnly={editMode}
+            maxAccess={maxAccess}
+            onClear={() => formik.setFieldValue('codeId', '')}
```
Comment (NoraBitarSM) [inline src/pages/strategies/forms/strategiesForm.js:135]: do we need this or the resourceCombobox does the job ? 
same for the onChange event.

File: src/pages/strategies/forms/strategiesForm.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+          <ResourceComboBox
+            datasetId={DataSets.SY_TYPE}
+            name='type'
+            label={labels.type}
+            valueField='key'
+            displayField='value'
+            values={strategiesFormik.values}
+            required
+            maxAccess={maxAccess}
+            onChange={(event, newValue) => {
```
Comment (NoraBitarSM) [inline src/pages/strategies/forms/strategiesForm.js:152]: do we need this ?

File: src/pages/strategies/windows/strategiesWindow.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+          labels={labels}
+          setEditMode={setEditMode}
+          setStore={setStore}
+          maxAccess={maxAccess}
+          store={store}
+          strategiesFormik={strategiesFormik}
+        />
+      </CustomTabPanel>
+      <CustomTabPanel height={height} index={2} value={activeTab}>
+        <PereList labels={labels} setEditMode={setEditMode} setStore={setStore} maxAccess={maxAccess} store={store} />
```
Comment (NoraBitarSM) [inline src/pages/strategies/windows/strategiesWindow.js:56]: store for that ? 
check the tabs. are you sure you need to send setStore and store props ?

**Review by NoraBitarSM — CHANGES_REQUESTED** [review]

File: src/pages/strategies/forms/indicatorForm.js [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    }
+  })
+
+  const submitIndicators = obj => {
+    if (obj.length === 0) {
+      return
+    }
+
+    const requests = obj.map(item => {
+      const { id, indicatorRef, indicatorName, ...itemWithoutId } = item
```
[the remark this answers is not in the document]
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/indicatorForm.js:69]: sending the data to the endpoint without id indicatorRef indicator name

File: src/pages/strategies/forms/indicatorForm.js [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        console.error(error)
+        setErrorMessage(error)
+      })
+  }
+
+  useEffect(() => {
+    if (recordId) {
+      getValueGridData(recordId)
+    }
+  }, [recordId, applyTrigger])
```
[the remark this answers is not in the document]
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/indicatorForm.js:115]: the applyTrigger when the apply button is pressed the useEffect give the initial values of the get request

[the remark this answers is not in the document]
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/indicatorForm.js:115]: the apply trigger is used to get the value from the .qry on apply button the apply button change the use effect  and get the values on click

File: src/pages/strategies/forms/strategiesForm.js (hunk 5 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/strategiesForm.js:135]: no we dont need the onClear  i remove it 
but i need the onChange

File: src/pages/strategies/forms/strategiesForm.js (hunk 6 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/strategiesForm.js:152]: yes we need it

File: src/pages/strategies/forms/indicatorForm.js (hunk 1 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/indicatorForm.js:100]: yes we can i will change it

File: src/pages/dr-groups/forms/ApproverList.js [hunk 10]
```diff
@@ -29,42 +29,34 @@ import { useInvalidate, useResourceQuery } from 'src/hooks/resource'
...
       })
   }
-  useEffect(()=>{
+  useEffect(() => {
     recordId && getValueGridData(recordId)
-    
-  },[recordId,refresh])
+  }, [recordId, refresh])
 
   // const {
```
Comment (omarZrayka23) [inline src/pages/dr-groups/forms/ApproverList.js:61]: can remove it?

File: src/pages/dr-groups/forms/ApproverList.js [hunk 11]
```diff
@@ -76,87 +68,71 @@ const ApproverList = ({store,labels,maxAccess}) => {
...
   }
 
-  function openForm2 (recordId){
+  function openForm2(recordId) {
     stack({
       Component: ApproverForm,
       props: {
         labels: labels,
-        recordId: recordId? recordId : null,
+        recordId: recordId ? recordId : null,
```
Comment (omarZrayka23) [inline src/pages/dr-groups/forms/ApproverList.js:103]: recordId: recordId ?? null 
u can do it  same result

File: src/pages/strategies/forms/codeList.js [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+      .then(res => {
+        setValueGridData(res)
+      })
+      .catch(error => {})
+  }
+  useEffect(() => {
+    recordId && getValueGridData(recordId)
+  }, [recordId, refresh])
+
+  // const {
```
Comment (omarZrayka23) [inline src/pages/strategies/forms/codeList.js:49]: remove comment code

File: src/pages/strategies/forms/indicatorForm.js [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+    {
+      component: 'resourcecombobox',
+      label: labels.indicator,
+
+      name: 'indicatorName',
+      props: {
+        endpointId: DocumentReleaseRepository.ReleaseIndicator.qry,
+        valueField: 'recordId',
+        displayField: 'name',
+        parameters: `_startAt=0&_pageSize=50`,
```
Comment (omarZrayka23) [inline src/pages/strategies/forms/indicatorForm.js:152]: put parameters under endpointId direct

File: src/pages/strategies/forms/indicatorForm.js [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+        mapping: [
+          { from: 'name', to: 'indicatorName' },
+          { from: 'reference', to: 'indicatorRef' },
+
```
Comment (omarZrayka23) [inline src/pages/strategies/forms/indicatorForm.js:156]: remove space

**Review by hadichahine — CHANGES_REQUESTED** [review]

File: src/pages/dr-groups/forms/ApproverForm.js [hunk 15]
```diff
@@ -7,153 +6,118 @@ import { ResourceIds } from 'src/resources/ResourceIds'
...
-  
-    const { postRequest, getRequest} = useContext(RequestsContext)
-    
-    const {recordId:grId} = store
+  const [initialValues, setInitialData] = useState({
+    codeId: '',
+    groupId: grId
+  })
 
+  const formik = useFormik({
```
[the remark this answers is not in the document]
Comment (Alinadimmokdad) [reply src/pages/dr-groups/forms/ApproverForm.js]: okay i will

File: src/pages/dr-groups/forms/ApproverForm.js [hunk 16]
```diff
@@ -7,153 +6,118 @@ import { ResourceIds } from 'src/resources/ResourceIds'
...
-    recordId,
-    store,
-    setRefresh,
-    
-  }) => {
-  
-    const { postRequest, getRequest} = useContext(RequestsContext)
-    
-    const {recordId:grId} = store
+  const [initialValues, setInitialData] = useState({
```
[the remark this answers is not in the document]
Comment (Alinadimmokdad) [reply src/pages/dr-groups/forms/ApproverForm.js]: i did it with the useForm

File: src/pages/dr-groups/forms/ApproverForm.js [hunk 17]
```diff
@@ -7,153 +6,118 @@ import { ResourceIds } from 'src/resources/ResourceIds'
...
+      if (isNewRecord) {
+        toast.success('Record Added Successfully')
+        setInitialData(prevData => ({
+          ...prevData,
+          ...obj
+        }))
+        setEditMode(true)
+      } else {
+        toast.success('Record Edited Successfully')
+        setInitialData(prevData => ({
```
[the remark this answers is not in the document]
Comment (Alinadimmokdad) [reply src/pages/dr-groups/forms/ApproverForm.js]: done

File: src/pages/dr-groups/forms/ApproverForm.js [hunk 18]
```diff
@@ -7,153 +6,118 @@ import { ResourceIds } from 'src/resources/ResourceIds'
...
-      }),
-      onSubmit: values => {
-        postGroups(values)
-    
+          ...obj
+        }))
+        console.log('ooooooooooooooooooooooooooo', obj.codeId)
       }
+      setRefresh(prev => !prev)
+    } catch (error) {
```
[the remark this answers is not in the document]
Comment (Alinadimmokdad) [reply src/pages/dr-groups/forms/ApproverForm.js]: done

File: src/pages/dr-groups/forms/ApproverList.js [hunk 19]
```diff
@@ -22,141 +22,100 @@ import { RequestsContext } from 'src/providers/RequestsContext'
...
   const { getRequest, postRequest } = useContext(RequestsContext)
-  const {recordId}= store
+  const { recordId } = store
   const [selectedRecordId, setSelectedRecordId] = useState(null)
   const { stack } = useWindow()
-  const [valueGridData , setValueGridData] = useState()
+  const [valueGridData, setValueGridData] = useState()
 
-  const [refresh,setRefresh]=useState(false)
+  const [refresh, setRefresh] = useState(false)
```
[the remark this answers is not in the document]
Comment (Alinadimmokdad) [reply src/pages/dr-groups/forms/ApproverList.js:14]: done

File: src/pages/strategies/forms/strategiesForm.js [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+
+  const [initialValues, setInitialData] = useState({
+    recordId: null,
+    name: '',
+
+    type: '',
+    groupId: ''
+  })
+
+  const strategiesFormik = useFormik({
```
Comment (hadichahine) [inline src/pages/strategies/forms/strategiesForm.js:35]: Should use `useForm`?

File: src/pages/strategies/forms/strategiesForm.js [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+      type: yup.string().required()
+    }),
+    onSubmit: values => {
+      postGroups(values)
+    }
+  })
+
+  const prevValuesRef = useRef()
+
+  useEffect(() => {
```
Comment (hadichahine) [inline src/pages/strategies/forms/strategiesForm.js:51]: Why do we need this? I don't like the prevValues, newValues comparison on this form.

File: src/pages/strategies/forms/PrereqList.js [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+  const addCode = () => {
+    openForm()
+  }
+
+  const delCode = async obj => {
+    await postRequest({
+      extension: DocumentReleaseRepository.StrategyPrereq.del,
+      record: JSON.stringify(obj)
+    })
+    setRefresh(prev => !prev)
```
Comment (hadichahine) [inline src/pages/strategies/forms/PrereqList.js:70]: Why do we still need this?

File: src/pages/strategies/forms/PrereqForm.js [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+    prerequisiteId: '',
+    StrategyId: stgId
+  })
+
+  const validationSchema = yup.object({
+    codeId: yup.number().required(),
+    prerequisiteId: yup.number().required()
+  })
+
+  const formik = useFormik({
```
Comment (hadichahine) [inline src/pages/strategies/forms/PrereqForm.js:34]: useForm?

File: src/pages/strategies/forms/PrereqForm.js [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+        setEditMode(true)
+      } else {
+        toast.success('Record Edited Successfully')
+        setInitialData(prevData => ({
+          ...prevData,
+
+          ...obj
+        }))
+      }
+      setRefresh(prev => !prev)
```
Comment (hadichahine) [inline src/pages/strategies/forms/PrereqForm.js:68]: Set refresh?

File: src/pages/strategies/forms/PrereqForm.js [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+
+    try {
+      const res = await postRequest({
+        extension: DocumentReleaseRepository.StrategyPrereq.set,
+        record: JSON.stringify(obj)
+      })
+
+      if (isNewRecord) {
+        toast.success('Record Added Successfully')
+        setInitialData(prevData => ({
```
Comment (hadichahine) [inline src/pages/strategies/forms/PrereqForm.js:55]: why do we need to setInitialData on submit?

File: src/pages/strategies/forms/CodeForm.js [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+      } else {
+        toast.success('Record Edited Successfully')
+
+        setInitialData(prevData => ({
+          ...prevData,
+
+          ...obj
+        }))
+      }
+      setRefresh(prev => !prev)
```
Comment (hadichahine) [inline src/pages/strategies/forms/CodeForm.js:81]: Same here.

File: src/pages/strategies/forms/CodeForm.js [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+
+    try {
+      const res = await postRequest({
+        extension: DocumentReleaseRepository.StrategyCode.set,
+        record: JSON.stringify(obj)
+      })
+
+      if (isNewRecord) {
+        toast.success('Record Added Successfully')
+        setInitialData(prevData => ({
```
Comment (hadichahine) [inline src/pages/strategies/forms/CodeForm.js:67]: Setting initial data.

File: src/pages/strategies/forms/CodeForm.js [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+    validationSchema: yup.object({
+      codeId: yup.string().required()
+    }),
+    onSubmit: async (values, { setSubmitting }) => {
+      try {
+        await postGroups(values)
+        window.close()
+      } catch (error) {
+        toast.error('An error occurred')
+        setSubmitting(false)
```
Comment (hadichahine) [inline src/pages/strategies/forms/CodeForm.js:51]: Why do we need this? Was this code copied from somewhere in the codebase or chatgpt'd?

File: src/pages/strategies/forms/codeList.js [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+  const addCode = () => {
+    openForm()
+  }
+
+  const delCode = async obj => {
+    await postRequest({
+      extension: DocumentReleaseRepository.StrategyCode.del,
+      record: JSON.stringify(obj)
+    })
+    setRefresh(prev => !prev)
```
Comment (hadichahine) [inline src/pages/strategies/forms/codeList.js:71]: same here

File: src/pages/strategies/forms/codeList.js [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+  const { getRequest, postRequest } = useContext(RequestsContext)
+  const { recordId } = store
+  const [selectedRecordId, setSelectedRecordId] = useState(null)
+  const { stack } = useWindow()
+  const [valueGridData, setValueGridData] = useState()
+
+  const [refresh, setRefresh] = useState(false)
+
+  //states
+  const [windowOpen, setWindowOpen] = useState(false)
```
Comment (hadichahine) [inline src/pages/strategies/forms/codeList.js:32]: We're not using those?

File: src/pages/strategies/forms/codeList.js [hunk 31]
```diff
@@ -0,0 +1,117 @@
...
+  const [errorMessage, setErrorMessage] = useState(null)
```
Comment (hadichahine) [inline src/pages/strategies/forms/codeList.js:33]: same here? please omit state that we're not using.

File: src/pages/strategies/forms/indicatorForm.js [hunk 32]
```diff
@@ -0,0 +1,201 @@
...
+  editMode,
+  height,
+  maxAccess,
+  store,
+  onApply,
+  setStore
+}) => {
+  const [valueGridData, setValueGridData] = useState([])
+  const { postRequest } = useContext(RequestsContext)
+  const [refresh, setRefresh] = useState(false)
```
Comment (hadichahine) [inline src/pages/strategies/forms/indicatorForm.js:26]: remove

File: src/pages/strategies/forms/indicatorForm.js [hunk 33]
```diff
@@ -0,0 +1,201 @@
...
+  const { getRequest } = useContext(RequestsContext)
+  const { recordId } = store
+
+  const [applyTrigger, setApplyTrigger] = useState(0)
+
+  const formik = useFormik({
```
Comment (hadichahine) [inline src/pages/strategies/forms/indicatorForm.js:32]: useForm

**Review by hadichahine — CHANGES_REQUESTED** [review]
Please make sure that any state that's not being used is omitted.

File: src/components/Shared/Buttons.js [hunk 34]
```diff
@@ -87,12 +87,9 @@ export const Buttons = [
   },
   {
     key: 'Apply',
```
[the remark this answers is not in the document]
Comment (Alinadimmokdad) [reply src/components/Shared/Buttons.js:89]: apply button

File: src/pages/strategies/forms/CodeForm.js (hunk 28 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/CodeForm.js:51]: the code Form should be closed after submit

File: src/pages/strategies/forms/CodeForm.js (hunk 26 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/CodeForm.js:81]: done

File: src/pages/strategies/forms/PrereqForm.js (hunk 23 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/PrereqForm.js:34]: okay

File: src/pages/strategies/forms/PrereqForm.js (hunk 25 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/PrereqForm.js:55]: we dont

File: src/pages/strategies/forms/PrereqForm.js (hunk 24 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/PrereqForm.js:68]: i used refetch now

File: src/pages/strategies/forms/PrereqList.js (hunk 22 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/PrereqList.js:70]: i will use refetch with invalidate()

File: src/pages/strategies/forms/codeList.js (hunk 30 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/codeList.js:32]: no i will remove

File: src/pages/strategies/forms/codeList.js (hunk 29 above)
Comment (Alinadimmokdad) [reply src/pages/strategies/forms/codeList.js:71]: done

**Pull request merged**